gluonnlp/model/sampled_block.py

```diff
@@ -0,0 +1,623 @@
+    num_sampled: int
+        Number of classes randomly sampled for each batch.
+    in_unit: int
+        Dimensionality of the input space.
+    remove_accidental_hits: bool
+        Whether to remove "accidental hits" when a sampled candidate is equal to
+        one of the true classes.
+    """
+    def __init__(self, num_classes, num_sampled, in_unit, remove_accidental_hits,
+                 prefix=None, params=None):
+        super(_SampledLogitsHelper, self).__init__(prefix=prefix, params=params)
+        self._num_classes = num_classes
+        self._num_sampled = num_sampled
+        self._in_unit = in_unit
+        self._remove_accidental_hits = remove_accidental_hits
+
+    def hybrid_forward(self, F, x, sampled_candidates, expected_count_sampled,
+                       expected_count_true, label, w_all, b_all):
+        """Forward computation."""
+        # (num_sampled, in_unit)
+        w_sampled = w_all.slice(begin=(0, 0), end=(self._num_sampled, None))
+        w_true = w_all.slice(begin=(self._num_sampled, 0), end=(None, None))
+        b_sampled = b_all.slice(begin=(0,), end=(self._num_sampled,))
+        b_true = b_all.slice(begin=(self._num_sampled,), end=(None,))
+        # true logits
+        # (batch_size, 1)
+        x = x.reshape((-1, self._in_unit))
+        logits_true = (w_true * x).sum(axis=1) + b_true
+        # samples logits
+        # (batch_size, num_sampled)
+        b_sampled = F.reshape(b_sampled, (-1,))
+        logits_sampled = F.FullyConnected(x, weight=w_sampled, bias=b_sampled,
+                                          num_hidden=self._num_sampled)
+
+        # remove accidental hits
+        if self._remove_accidental_hits:
+            label_vec = F.reshape(label, (-1, 1))
+            sample_vec = F.reshape(sampled_candidates, (1, -1))
+            mask = F.broadcast_equal(label_vec, sample_vec) * -1e37
+            logits_sampled = logits_sampled + mask
+
+        # subtract log(q)
+        expected_count_sampled = F.reshape(expected_count_sampled,
+                                           shape=(1, self._num_sampled))
+        expected_count_true = expected_count_true.reshape((-1,))
+        logits_true = logits_true - F.log(expected_count_true)
+        logits_true = logits_true.reshape((-1, 1))
+        logits_sampled = F.broadcast_sub(logits_sampled, F.log(expected_count_sampled))
+
+        # logits and new_labels
+        # (batch_size, 1+num_sampled)
+        logits = F.concat(logits_true, logits_sampled, dim=1)
+        return logits, F.zeros_like(label)
+
+    def __repr__(self):
+        s = '{name}({mapping})'
+        mapping = '{0} -> {1}, with {2} samples'.format(self._in_unit, self._num_classes,
+                                                        self._num_sampled)
+        return s.format(name=self.__class__.__name__,
+                        mapping=mapping,
+                        **self.__dict__)
+
+class _SampledLogits(HybridBlock):
```

Comment:
Can `_SampledLogits` and `_SparseSampledLogits` be merged with a flag representing whether it is Sparse or not? Since the `init` and `forward` share the same logic. So based on this, can the `SparseNCELogits` and `NCELogits` be merged? Similarly for `ISLogits` and `SparseISLogits`?

Reply:
I thought about that but one of them is HybridBlock and the other is not. @szha can python dynamically change inheritance relationships?

Reply:
no. let's keep it separate. having divergence in implementation can be hard to maintain.

Reply:
Well, they should be separated since 1) they share same logic, 2) sparseLogits (the reason why Logit computation is separate from the softmax computation) is designed for sparse or large data to approximate the full softmax. One can simply use softmax layer to compute the full softmax if the data size is small.

Reply:
these are certainly valid reasons. practically, it would likely be easier to maintain, at least until gluon can better mix different block types. so from engineering perspective, I'd recommend that we keep it separate for now, keep a note, and revisit this decision once our dependency is in a better state. is that ok, @cgraywang ?

Reply:
It looks this is the solution for now. Thanks @szha